src/python/BuildingControlsSimulator/DataClients/WeatherChannel.py

```diff
@@ -703,3 +734,303 @@ def to_epw(
+            time_index = pd.to_datetime(f['time_index'][...].astype(str), utc=True)# Temporal resolution is 30min
+
+            #get data sets
+            dset_dni  = f['dni']
+            dset_ghi  = f['ghi']
+            dset_temp = f['air_temperature']
+            dset_rh   = f['relative_humidity']
+            # list(f)  # for full list the datasets in the file
+
+            #filter on time series for given location
+            tseries_dni  = dset_dni[:,  location_idx] / dset_dni.attrs['psm_scale_factor']  
+            tseries_ghi  = dset_ghi[:,  location_idx] / dset_ghi.attrs['psm_scale_factor']  
+            tseries_temp = dset_temp[:, location_idx] / dset_temp.attrs['psm_scale_factor'] 
+            tseries_rh   = dset_rh[:,   location_idx] / dset_rh.attrs['psm_scale_factor']   
+
+            #combine into single DF int
+            df_solar = pd.DataFrame()
+            df_solar["datetime"]          = time_index   # datetime64
+            df_solar["dni"]               = tseries_dni  # W/m2, float
+            df_solar["ghi"]               = tseries_ghi  # W/m2, float
+            df_solar["temp_air"]          = tseries_temp # degC, float
+            df_solar["relative_humidity"] = tseries_rh   # %rh,  float
+
+            df_solar.to_csv(strPath + strFile, index=False, compression='gzip')
+        else:
+            print('Re-opening data')
+            df_solar = pd.read_csv(strFile, compression='gzip')
+            df_solar.datetime = pd.to_datetime(df_solar.datetime)
+        
+        return strFile, df_solar
+    # test:
+    # strFile, df_solar = get_nsrdb(51.1739243, -114.1643148)
+    # df_solar.info()
+
+
+    fill_nsrdb_data = attr.ib(default=None)
+    def fill_nsrdb(self, input_epw_data, datetime_channel, fill_nsrdb_data, sim_config):
+        """Fill input data with NSRDB 2019 data as available.
+        All data is internally in UTC.
+
+        :param epw_data: EnergyPlus Weather data in a dataframe of epw_columns
+        :type epw_data: pd.DataFrame
+
+        :param datetime_column: datetime column in fill data.
+        "type datetime_column: str
+        """
+        if input_epw_data.empty:
+            input_epw_data.columns = self.epw_columns
+            return input_epw_data
+
+        if fill_nsrdb_data.empty:
+            raise ValueError(f"fill_epw_data is empty.")
+
+        # save fill_nsrdb_data that was actually used to fill
+        self.fill_nsrdb_data = fill_nsrdb_data
+
+        # edit unique copy of input df
+        epw_data = input_epw_data.copy(deep=True)
+
+        # add datetime column for merge with fill data
+        epw_data = pd.concat(
+            [
+                datetime_channel.data[datetime_channel.spec.datetime_column],
+                epw_data,
+            ],
+            axis="columns",
+        ).rename(columns={datetime_channel.spec.datetime_column: self.datetime_column})
+
+        # # using 2019 NSRDB data there may be missing rows at beginning
```

Comment:
Decoupling from epw generation will fix this problem.

Reply:
done.